File: bytebrew-cli/src/infrastructure/mobile/EventBroadcaster.ts
// EventBroadcaster - subscribes to IEventBus, converts DomainEvents to MobileMessages,
// and broadcasts to subscribed mobile devices via BridgeMessageRouter.

import { v4 as uuidv4 } from 'uuid';
import { getLogger, type Logger } from '../../lib/logger.js';
import type {
  IEventBus,
  DomainEvent,
} from '../../domain/ports/IEventBus.js';
import type { IBridgeMessageRouter, MobileMessage } from '../bridge/BridgeMessageRouter.js';
import type { IEventBuffer } from './EventBuffer.js';

/** Serialized event payload sent to mobile clients */
export interface SerializedEvent {
  type: string;
  [key: string]: unknown;
}

/** Subscription entry: a device subscribed to a specific session's events */
interface DeviceSubscription {
  deviceId: string;
  sessionId: string | undefined; // undefined = all sessions
}

/**
 * Subscribes to the CLI EventBus, converts DomainEvents into MobileMessage
 * format, and sends them to all subscribed mobile devices through the
 * BridgeMessageRouter. Buffers events via IEventBuffer for reconnect backfill.
 */
export class EventBroadcaster {
  private readonly logger: Logger;
  private readonly subscriptions = new Map<string, DeviceSubscription>(); // deviceId -> subscription
  private unsubscribeEventBus: (() => void) | null = null;

  constructor(
    private readonly eventBus: IEventBus,
    private readonly router: IBridgeMessageRouter,
    private readonly eventBuffer: IEventBuffer<SerializedEvent>,
    private readonly sessionId: string,
  ) {
    this.logger = getLogger().child({ component: 'EventBroadcaster' });
  }

  /** Subscribe to all domain events and start broadcasting */
  start(): void {
    if (this.unsubscribeEventBus) {
      this.logger.warn('EventBroadcaster already started');
      return;
    }

    this.unsubscribeEventBus = this.eventBus.subscribeAll((event) => {
      this.handleDomainEvent(event);
    });

    this.logger.info('EventBroadcaster started');
  }

  /** Unsubscribe from all domain events */
  stop(): void {
    if (this.unsubscribeEventBus) {
      this.unsubscribeEventBus();
      this.unsubscribeEventBus = null;
    }

    this.subscriptions.clear();
    this.logger.info('EventBroadcaster stopped');
  }

  /** Subscribe a mobile device to receive events */
  subscribe(deviceId: string, sessionId?: string): void {
    this.subscriptions.set(deviceId, { deviceId, sessionId });
    this.logger.info('Device subscribed', { deviceId, sessionId });
  }

  /** Unsubscribe a mobile device */
  unsubscribe(deviceId: string): void {
    if (this.subscriptions.delete(deviceId)) {
      this.logger.info('Device unsubscribed', { deviceId });
    }
  }

  /** Get buffered events for backfill on reconnect */
  getBufferedEvents(
    sessionId: string,
    afterIndex: number,
  ): { events: SerializedEvent[]; lastIndex: number } {
    return this.eventBuffer.getAfter(sessionId, afterIndex);
  }

  // --- Private ---

  private handleDomainEvent(event: DomainEvent): void {
    const serialized = serializeEvent(event);

    // Buffer the event for potential backfill
    this.eventBuffer.push(this.sessionId, serialized);

    // Broadcast to all subscribed devices
    if (this.subscriptions.size === 0) return;

    for (const [deviceId, sub] of this.subscriptions) {
      // If device subscribed to a specific session, filter by it
      if (sub.sessionId && sub.sessionId !== this.sessionId) {
        continue;
      }

      const message: MobileMessage = {
        type: 'session_event',
        request_id: uuidv4(),
        device_id: deviceId,
        payload: {
          session_id: this.sessionId,
          event: serialized,
        },
      };

      try {
        this.router.sendMessage(deviceId, message);
      } catch (err) {
        const errorMessage = err instanceof Error ? err.message : String(err);
        this.logger.error('Failed to broadcast event to device', {
          deviceId,
          eventType: event.type,
          error: errorMessage,
        });
      }
    }
  }
}

/**
 * Converts a DomainEvent into a serializable plain object for mobile clients.
 * Mirrors the mapping from MobileProxyServer.serializeEvent.
 */
function serializeEvent(event: DomainEvent): SerializedEvent {
  switch (event.type) {
    case 'MessageCompleted':
      return {
        type: event.type,
        message: event.message.toSnapshot(),
      };

    case 'MessageStarted':
      return {
        type: event.type,
        messageId: event.messageId,
        role: event.role,
      };

    case 'ToolExecutionStarted':
    case 'ToolExecutionCompleted':
      return {
        type: event.type,
        execution: event.execution.toSnapshot(),
      };

    case 'AskUserRequested':
      return {
        type: event.type,
        questions: event.questions,
      };

    case 'StreamingProgress':
      return {
        type: event.type,
        messageId: event.messageId,
        tokensAdded: event.tokensAdded,
        totalTokens: event.totalTokens,
      };

    case 'ProcessingStarted':
    case 'ProcessingStopped':
      return { type: event.type };

    case 'ErrorOccurred':
      return {
        type: event.type,
        message: event.error.message,
        context: event.context,
      };

    case 'AgentLifecycle':
      return {
        type: event.type,
        lifecycleType: event.lifecycleType,
        agentId: event.agentId,
        description: event.description,
      };

    case 'AskUserResolved':
      return { type: event.type };
  }
}

File: bytebrew-srv/internal/infrastructure/bridge/event_broadcaster.go
package bridge

import (
	"log/slog"
	"sync"

	"github.com/google/uuid"
	pb "github.com/syntheticinc/bytebrew/bytebrew-srv/api/proto/gen"
)

// DeviceSubscription tracks a mobile device's subscription to a session.
type DeviceSubscription struct {
	DeviceID    string
	SessionID   string
	LastEventID string
}

// MessageSender sends a message to a specific device.
type MessageSender interface {
	SendMessage(deviceID string, msg *MobileMessage) error
}

// EventBroadcaster serializes SessionEvents into the flat mobile format,
// buffers them for reconnect backfill, and sends them to subscribed devices
// via the MessageSender.
type EventBroadcaster struct {
	sender MessageSender
	buffer *EventBuffer

	subscribers map[string]*DeviceSubscription // deviceID → subscription
	mu          sync.RWMutex
}

// NewEventBroadcaster creates a new EventBroadcaster.
func NewEventBroadcaster(sender MessageSender) *EventBroadcaster {
	return &EventBroadcaster{
		sender:      sender,
		buffer:      NewEventBuffer(1000),
		subscribers: make(map[string]*DeviceSubscription),
	}
}

// Subscribe registers a device to receive events for the given session.
// If lastEventID is provided, missed events are backfilled immediately.
func (b *EventBroadcaster) Subscribe(deviceID, sessionID, lastEventID string) {
	b.mu.Lock()
	b.subscribers[deviceID] = &DeviceSubscription{
		DeviceID:    deviceID,
		SessionID:   sessionID,
		LastEventID: lastEventID,
	}
	b.mu.Unlock()

	slog.Info("device subscribed to session", "device_id", deviceID, "session_id", sessionID)

	if lastEventID == "" {
		return
	}

	// Backfill missed events.
	missed := b.buffer.GetAfter(lastEventID)
	for _, evt := range missed {
		if evt.SessionID != sessionID {
			continue
		}
		b.sendToDevice(deviceID, sessionID, evt.Event, evt.EventID)
	}
}

// Unsubscribe removes a device's subscription.
func (b *EventBroadcaster) Unsubscribe(deviceID string) {
	b.mu.Lock()
	defer b.mu.Unlock()

	delete(b.subscribers, deviceID)
	slog.Info("device unsubscribed", "device_id", deviceID)
}

// BroadcastEvent serializes a SessionEvent to the flat mobile format, buffers
// it, and fans out to all devices subscribed to the event's session.
func (b *EventBroadcaster) BroadcastEvent(sessionID string, event *pb.SessionEvent) {
	serialized := serializeEventForMobile(event)
	if serialized == nil {
		return
	}

	eventID := b.buffer.Append(sessionID, serialized)

	b.mu.RLock()
	var targets []*DeviceSubscription
	for _, sub := range b.subscribers {
		if sub.SessionID == sessionID {
			targets = append(targets, sub)
		}
	}
	b.mu.RUnlock()

	for _, sub := range targets {
		b.sendToDevice(sub.DeviceID, sessionID, serialized, eventID)
	}
}

// SendSessionStatus sends a synthetic session status event to a specific device.
// Used after subscribe to ensure the device knows the current processing state,
// preventing stuck-spinner when ProcessingStopped was lost during TCP death.
//
// NOT buffered: synthetic events are sent without event_id so the mobile client's
// dedup logic (based on event_id) won't skip them. This avoids collisions when
// the server restarts and the mevt counter resets (mobile may already have old
// mevt-1 in its seen set).
func (b *EventBroadcaster) SendSessionStatus(deviceID, sessionID string, processing bool) {
	eventType := "ProcessingStopped"
	state := "idle"
	if processing {
		eventType = "ProcessingStarted"
		state = "processing"
	}

	statusEvent := map[string]interface{}{
		"type":  eventType,
		"state": state,
	}
	// Empty event_id → mobile skips dedup check → always processed.
	b.sendToDevice(deviceID, sessionID, statusEvent, "")
}

func (b *EventBroadcaster) sendToDevice(deviceID, sessionID string, event map[string]interface{}, eventID string) {
	msg := &MobileMessage{
		Type:      "session_event",
		RequestID: uuid.New().String(),
		DeviceID:  deviceID,
		Payload: map[string]interface{}{
			"session_id": sessionID,
			"event":      event,
			"event_id":   eventID,
		},
	}

	if err := b.sender.SendMessage(deviceID, msg); err != nil {
		slog.Error("broadcast to device failed", "device_id", deviceID, "event_id", eventID, "error", err)
	}
}

// serializeEventForMobile converts a proto SessionEvent into the flat JSON
// format expected by mobile clients.
func serializeEventForMobile(event *pb.SessionEvent) map[string]interface{} {
	switch event.GetType() {
	case pb.SessionEventType_SESSION_EVENT_ANSWER:
		return map[string]interface{}{
			"type":     "MessageCompleted",
			"content":  event.GetContent(),
			"role":     "assistant",
			"agent_id": event.GetAgentId(),
		}

	case pb.SessionEventType_SESSION_EVENT_ANSWER_CHUNK:
		return map[string]interface{}{
			"type":     "StreamingProgress",
			"content":  event.GetContent(),
			"agent_id": event.GetAgentId(),
		}

	case pb.SessionEventType_SESSION_EVENT_TOOL_EXECUTION_START:
		args := make(map[string]interface{}, len(event.GetToolArguments()))
		for k, v := range event.GetToolArguments() {
			args[k] = v
		}
		return map[string]interface{}{
			"type":      "ToolExecutionStarted",
			"call_id":   event.GetCallId(),
			"tool_name": event.GetToolName(),
			"arguments": args,
			"agent_id":  event.GetAgentId(),
		}

	case pb.SessionEventType_SESSION_EVENT_TOOL_EXECUTION_END:
		return map[string]interface{}{
			"type":           "ToolExecutionCompleted",
			"call_id":        event.GetCallId(),
			"tool_name":      event.GetToolName(),
			"result_summary": event.GetToolResultSummary(),
			"has_error":      event.GetToolHasError(),
			"agent_id":       event.GetAgentId(),
		}

	case pb.SessionEventType_SESSION_EVENT_REASONING:
		return map[string]interface{}{
			"type":     "ReasoningChunk",
			"content":  event.GetContent(),
			"agent_id": event.GetAgentId(),
		}

	case pb.SessionEventType_SESSION_EVENT_ASK_USER:
		return map[string]interface{}{
			"type":     "AskUserRequested",
			"question": event.GetQuestion(),
			"options":  event.GetOptions(),
			"agent_id": event.GetAgentId(),
		}

	case pb.SessionEventType_SESSION_EVENT_PROCESSING_STARTED:
		return map[string]interface{}{
			"type":  "ProcessingStarted",
			"state": "processing",
		}

	case pb.SessionEventType_SESSION_EVENT_PROCESSING_STOPPED:
		return map[string]interface{}{
			"type":  "ProcessingStopped",
			"state": "idle",
		}

	case pb.SessionEventType_SESSION_EVENT_ERROR:
		msg := event.GetContent()
		if detail := event.GetErrorDetail(); detail != nil {
			msg = detail.GetMessage()
		}
		return map[string]interface{}{
			"type":    "Error",
			"message": msg,
			"code":    "error",
		}

	case pb.SessionEventType_SESSION_EVENT_PLAN_UPDATE:
		steps := make([]map[string]interface{}, 0, len(event.GetPlanSteps()))
		for _, s := range event.GetPlanSteps() {
			steps = append(steps, map[string]interface{}{
				"title":  s.GetTitle(),
				"status": s.GetStatus(),
			})
		}
		return map[string]interface{}{
			"type":      "PlanUpdated",
			"plan_name": event.GetPlanName(),
			"steps":     steps,
			"agent_id":  event.GetAgentId(),
		}

	default:
		slog.Warn("unknown session event type", "type", event.GetType().String())
		return nil
	}
}
